Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs

```diff
@@ -174,33 +170,33 @@ public async Task<IEnumerable<string>> ListTableNames(CancellationToken cancella
             return (await result.ToListAsync(cancellationToken)).Select(i => i.StringValue);
         }
 
-        internal async Task<AsyncQldbSession> GetSession()
+        internal async Task<AsyncQldbSession> GetSession(CancellationToken token)
         {
-            this.driverBase.LogSessionPoolState();
-
-            if (await this.driverBase.PoolPermits.WaitAsync(QldbDriverBase<AsyncQldbSession>.DefaultTimeoutInMs))
-            {
-                return this.driverBase.SessionPool.Count > 0 ? this.driverBase.SessionPool.Take() :
-                    await this.StartNewSession();
-            }
-            else
+            AsyncQldbSession session = this.driverBase.GetSessionFromPool();
+            if (session == null)
             {
-                this.driverBase.Logger.LogError(ExceptionMessages.SessionPoolEmpty);
-                throw new QldbDriverException(ExceptionMessages.SessionPoolEmpty);
+                session = await this.StartNewSession(token);
             }
+
+            return session;
         }
 
-        private async Task<AsyncQldbSession> StartNewSession()
+        private async Task<AsyncQldbSession> StartNewSession(CancellationToken token)
         {
             try
             {
                 Session session = await Session.StartSessionAsync(
                     this.driverBase.LedgerName,
                     this.driverBase.SessionClient,
-                    this.driverBase.Logger);
+                    this.driverBase.Logger,
+                    token);
                 this.driverBase.Logger.LogDebug("Creating new pooled session with ID {}.", session.SessionId);
                 return new AsyncQldbSession(session, this.driverBase.Logger);
             }
+            catch (OperationCanceledException oce)
+            {
+                throw new QldbTransactionException(QldbTransactionException.DefaultTransactionId, false, oce);
```

Comment:
What does this mean, we have a transaction issue on starting as session?

Reply:
OperationCanceledException is the exception thrown when someone uses a CancellationToken to stop their async operation.